lib/Sema/CSOptimizer.cpp

```diff
@@ -110,11 +195,47 @@ static bool isArithmeticOperator(ValueDecl *decl) {
   return decl->isOperator() && decl->getBaseIdentifier().isArithmeticOperator();
 }
 
+/// Generic choices are supported only if they are not complex enough
+/// that would they'd require solving to figure out whether they are a
+/// potential match or not.
+static bool isSupportedGenericOverloadChoice(ValueDecl *decl,
+                                             GenericFunctionType *choiceType) {
+  // Same type requirements cannot be handled because each
+  // candidate-parameter pair is (currently) considered in isolation.
+  if (llvm::any_of(choiceType->getRequirements(), [](const Requirement &req) {
+        switch (req.getKind()) {
+        case RequirementKind::SameType:
+        case RequirementKind::SameShape:
+          return true;
+
+        case RequirementKind::Conformance:
+        case RequirementKind::Superclass:
+        case RequirementKind::Layout:
+          return false;
+        }
+      }))
+    return false;
+
+  // If there are no same-type requirements, allow signatures
+  // that use only concrete types or generic parameters directly
+  // in their parameter positions i.e. `(T, Int)`.
+
+  auto *paramList = getParameterList(decl);
```

Comment:
Since you have a `GenericFunctionType` here, maybe just look at `choiceType->getParams()`?

Reply:
I think that would mean that I would have to look through self for instance members and I just didn’t want to mess with that…